### PR TITLE
Implement dispatch cancellation

### LINKED ISSUE #1162: Write the transport graph to the DB atomically
When the dispatcher first receives a lattice, it persists the transport graph to the datastore. However, the metadata for the lattice, nodes, and edges are written in [separate DB transactions](https://github.com/AgnostiqHQ/covalent/blob/6ecbd17361d93762a1acb63b82849dfdff035a6e/covalent/_results_manager/result.py#L884). If the Covalent server were to crash in the middle of a `Result.persist()` operation, the DB could be left in an inconsistent state, where only part of a lattice is written. 

To prevent this, we ought to perform all writes that logically belong together in a single transaction.

### LINKED ISSUE #1471: Cancelling dispatches: DB, data access layer, and Data Manager
As part of the [epic to support cancelling tasks and dispatches](https://github.com/AgnostiqHQ/covalent-staging/issues/662), this issue tracks the proposed changes to the DB schema, data access layer, and the core Data Manager.

### Handling sublattices (see linked design doc for more details)

Canceling an electron that is actually a sublattice is a slightly subtle matter as one needs to consider several cases depending on when the cancel request arrives:

1. If the electron has not yet been scheduled, don't run it.
2. If the sublattice graph is being built, interrupt the
graph-building job.
3. If the sublattice graph has been built but not yet assigned a
dispatch id, all tasks in the sublattice dispatch will be marked
cancelled. To implement this, `data_manager.make_dispatch` queries the DB to see if the parent electron has been marked cancelled, and if so, propagates that property to all
sublattice jobs. 
4. If the sublattice has been assigned a dispatch id `sub_dispatch_id`, cancel that dispatch (recursively invoke `dispatcher.cancel_dispatch(sub_dispatch_id)`)

*Remark*: In the third scenario, ****both query and sublattice dispatch creation are performed in **the same DB transaction to ensure atomicity** — if `cancel_requested` changes between the two steps, the sublattice dispatch could miss the cancel request.

## Acceptance criteria

**Database and data access layer**

- [ ] Implement [DB schema V14](https://drive.google.com/file/d/19962JaBySZxRpR0ja7IQ3n-DDhY9BzY4/view?usp=share_link):
    - In `_db/models.py`, add a "Jobs" table to track information pertaining to running a job in an executor. This will contain the following fields:
        - id (primary)
        - `cancel_requested`: boolean indicating whether the client has requested that the task be canceled. Initialize to `False`
        - `cancel_successful`: boolean
        - `job_handle` -- identifier for the job returned by the executor backend, such as an AWS Job ARN or a Slurm job id. This will be empty until the job is actually submitted by the executor plugin and the plugin persists the returned data.
    - The Electrons table adds a `job_id` column referring to the foreign key `jobs.id`
- [ ] Add a simple interface to the Jobs table in `_db/jobdb.py`:
    - `def txn_get_job_record(session: Session, job_id: int) -> Dict`
        - Queries the DB using the provided SQLAlchemy `Session`
        - Returns: a dictionary representation of the job record
        
        ```python
        {
              "job_id": job_record.id,
              "cancel_requested": job_record.cancel_requested,
              "cancel_successful": job_record.cancel_successful,
              "job_handle": job_record.job_handle,
        }
        ```
        
        - Raises: a `MissingJobRecord` exception if the primary key `job_id` does not exist in Jobs table.
        
        This method will be invoked within a broader `Session` context; see `write_result_to_db.txn_insert_electrons_data` below.
        
    - `def get_job_records(job_ids: List[int]) -> List[Dict]`
        - Invoke `txn_get_job_record` for each primary key.
        - Returns: a list of **dictionary representations** of the corresponding job records (same order).
        - Use the same DB session for all queries
    - `def update_job_record(session: Session, job_id: int, cancel_requested: bool = None, cancel_successful: bool = None, job_handle: str = None)`
        - Updates the corresponding row in the Jobs table using the provided SQLAlchemy `session`
    - `def update_job_records(kargs_list: List[Dict])`
        - Batch-updates the specified job records in a single DB session. Each entry of the list is a kwarg dictionary for `update_job_records`.
    - `def to_job_ids(dispatch_id: str, task_ids: List[int]) -> List[int]`
        - Utility function to map a list of node ids to their corresponding Job record primary keys.

- [ ] In `upsert.py` and `write_result_to_db.py`,    make use of **DB transactions** to avoid race conditions (see above remarks on handling sublattices) and also persist `cancel_requested` when creating each Electron record.
    - `write_result_to_db.py`:
        - Rename the existing `insert_lattices_data` to `txn_insert_lattices_data`, and add an SQLAlchemy Session as the first keyword argument:
        
        ```python
        def txn_insert_lattices_data(
        	session: Session,
        	dispatch_id: str,
        	electron_id: int,
          ...
        )
        ```
        
        Perform all DB operations using the provided session, but do not commit the changes.
        
        - Rename the existing `insert_electrons_data` to `txn_insert_electrons_data`, and add the following keyword parameters: SQLAlchemy Session as the first kwarg; `cancel_requested`. Signature:
            
            ```python
            def txn_insert_electrons_data(
            	session: Session,
            	parent_dispatch_id: str,
            	type: str,
            	...
            	call_after_filename: str,
            	cancel_requested: bool,
            	created_at: dt,
            	updated_at: dt,
            	...
            )
            ```
            
            This should perform the following operations using the provided DB session:
            
            - Insert a record in the Jobs table with the provided value for `cancel_requested`.
            - Call `session.flush()` to obtain the job record’s primary key `job_record.id`.
            - Insert a record in the Electrons table with the rest of the information supplied in the function arguments, together with the job record’s primary key.
        
        Do not commit the changes since these methods will be invoked within a broader `Session` context.
        
        - Rename the existing `insert_electron_dependency_data` to `txn_insert_electron_dependency_data`, and add an SQLAlchemy Session as the first kwarg:
            
            ```python
            def txn_insert_electron_dependency_data(
            	session: Session,
            	dispatch_id: str,
            	lattice: LatticeClass,
            )
            ```
            
            Perform the DB operations on the provided Session instead of creating a new Session. Do not commit the changes since these methods will be invoked within a broader `Session` context.
            
    - Make `upsert._lattice_data` and `upsert._electron_data` accept an SQLAlchemy Session. Perform all DB operations using the provided session. Do not commit the changes since these methods will be invoked within a broader `Session` context.
        - `def _lattice_data(session: Session, result: Result, electron_id: int = None)`
        - `def _electron_data(session: Session, result: Result, cancel_requested: bool = False)`
        
        The methods in `[upsert.py](http://upsert.py)` should be updated to use the **transactional** methods in `write_result_to_db.py` :
        
        - `txn_insert_lattices_data`
        - `txn_insert_electrons_data`
        - `txn_insert_electron_dependency_data`
    - Create public wrapper methods `upsert.lattice_data` and `upsert.electron_data` that invoke the respective internal methods on a new SQLAlchemy Session. Update all external references to the internal methods `_lattice_data` and `_electron_data` to these new public methods.
        - `def lattice_data(result: Result, electron_id: int = None)`
        - `def electron_data(result: Result, cancel_requested: bool = False)`
    - Rename `upsert_electron_dependency_data` to `txn_upsert_electron_dependency_data`; accept a session as a parameter.
    - Add a `upsert.persist_result` method to persist lattice, electron, and edge data in a **single transaction**.
        
        ```python
        def persist_result(result: Result, electron_id: int = None)
        ```
        
        - `electron_id` is the id of the parent electron if `result` is a sublattice dispatch. If this is not `None` , the `cancel_requested` property of the parent electron’s job should be propagated to all of the Electron records for `result`
    
    Update `update.persist()` to use this instead of the private methods in `upsert.py`.
    

**Data Manager** 

- [ ] Introduce a **********************Job Manager********************** module  `_core/data_modules/job_manager.py`:
    - `async def set_cancel_requested(dispatch_id: str, task_ids: List[int])`
        - Resolves each `task_id` to its corresponding `job_id`, (use `jobdb.to_job_ids`) and calls `jobdb.update_job_records()` with argument
        
        ```python
        [
        	{"job_id": job_id_1, cancel_requested: True},
        	{"job_id": job_id_2, cancel_requested: True},
          ...
        ]
        ```
        
    - `async def set_job_handle(dispatch_id: str, task_id: int, job_handle: str)`
        - Update the `job_handle` for the job corresponding to the electron `(dispatch_id, task_id)` . Use `jobdb.to_job_ids` to obtain the job id.
    - `async def set_cancel_result(dispatch_id: str, task_id: int, cancel_successful: bool)`
        - Update the `cancel_successful`field for the job corresponding to the specified electron.
    - `async def get_jobs_metadata(dispatch_id: str, task_ids: List[int]) -> List[Dict]`:
        - Returns: the job records (`jobdb.get_job_records`) corresponding to the provided list of tasks.

- [ ] Add and adjust unit tests. Feel free to crib some tests from [here](https://github.com/AgnostiqHQ/covalent/tree/cancel-dispatch-simplified/tests/covalent_dispatcher_tests/).

### LINKED ISSUE #1472: Cancelling dispatches: Executors and Runner
As part of the [epic to support cancelling tasks and dispatches](https://github.com/AgnostiqHQ/covalent-staging/issues/662), this issue tracks the proposed changes to the executors and Runner, including the mechanism for executor plugins to request services of Core Covalent.

###  Executor-Runner Communication protocol (see the linked design doc for more context)

Request: Executor -> Runner

`(action, body)`

Supported actions:

- `"get"`:
    - `body="cancel_requested"`
    - Retrieves the task’s `cancel_requested` property from the job manager
- `"put"`
    - `body=("job_handle", json_serialized_job_handle)`
    - Stores the job handle for the the task in persistent storage
- `"bye"`
    - `body=None`
    - Terminates the `watch()` task.

Response: Runner -> Executor

`(True/False, response)` where the first  element indicates success or failure

## Acceptance Criteria ##

**New exception type**: `TaskCancelledError`

- [ ] Define this in `covalent/_shared_files/exception.py`

********BaseExecutor********

- [ ] Business logic for communicating with the Runner:
    - `def _init_runtime(self, loop = None, cancel_pool: ThreadPoolExecutor = None)`:
        - This method is invoked by the Runner in `_run_task` to initialize runtime attributes of the executor instance.
        - Effect: creates the following instance attributes:
            - `self._send_queue: asyncio.Queue` for sending messages to the Runner
            - `self._recv_queue: queue.Queue` for receiving messages from the Runner
            - `self._loop` to store an event loop `loop` — this will be the Runner’s event loop
            - `self._cancel_pool` to store the `ThreadPoolExecutor` `cancel_pool`. This will be a dedicated threadpool maintained by the Runner exclusively for cancelling jobs.
    - `def _notify(self, action: str, body: str = None)`
        - This will be invoked from within the `execute()` method, which runs in a different thread from the main server thread.
        - Puts the message `(action, body)` in `self._send_queue`, return immediately (use `asyncio.Queue.put_nowait`).
        - Since the Runner will be listening to `self._send_queue` from a different thread, this method must use `self._loop.call_soon_threadsafe()` to ensure the listener wakes up.
        - In practice, `body` will be a JSON string.
    - `def _wait_for_response(self, timeout: int = 5)`
        - Blocks for up to `timeout` seconds for a response in `self._recv_queue`. and raise a `Empty` exception if no response is received. See [docs](https://docs.python.org/3/library/queue.html#queue.Queue.get) for `Queue.get()`. The response will be of the form `(status, body)`. where `status` is `True|False`.
        - Raise `RuntimeError` if `status` is False, indicating some sort of failure.
    - `def _notify_sync(self, action: str, body: str = None)`
        - Convenience method to call `self._notify` and then block the calling thread using `self._wait_for_response`
- [ ] Methods for consumption by executor plugin authors:
    - `def set_job_handle(self, job_handle)`
        - Saves a job handle (such as a slurm job id or Braket jobArn) to the database.
        - `job_handle` is required to be JSON-serializable — typically a string or int
        - Invokes `self._notify_sync()` with the following message
            - action: `"put"`
            - body: `("job_handle", json.dumps(job_handle))`
    - `def get_cancel_requested(self)`
        - Checks whether the user has requested to cancel the job being processed by the executor instance.
        - Invokes `self._notify_sync()` with the following message
            - action: `"get"`
            - body: `"cancel_requested"`
- [ ] Modifications to `execute()`:
    - The return value of `execute()` is now `(output, stdout, stderr, status)`, where `status` is the job’s terminal status. Its possible values are:
        - `RESULT_STATUS.COMPLETED`
        - `RESULT_STATUS.FAILED`
        - `RESULT_STATUS.CANCELLED`
    - In `execute()`, catch `TaskCancelledError` from `run()` and set the task’s terminal `job_status` to `RESULT_STATUS.CANCELLED`
    - The handler for `TaskRuntimeError` sets the terminal `job_status` to `RESULT_STATUS.FAILED`
    - Call `self._notify()` in the `finally:` block with the following message:
        - action: `"bye"`
        - body: `None`
        
        This tells the Runner to stop listening for messages from this executor instance.
        
- [ ] Methods to cancel jobs:
    - `def cancel(self, task_metadata: Dict, job_handle: Any) -> bool`
        - To be overridden by subclasses with backend-specific logic to cancel the job with the specified `job_handle`
        - Return `True` if job cancellation succeeded and `False` otherwise
        - Default implementation should simply return `False` and log a warning that job cancellation is not supported for the specific executor type.
    - `async def _cancel(self, task_metadata: Dict, job_handle: Any) -> bool`
        - To be invoked by the Runner
            - Calls `self.cancel()` followed by `self.teardown()` using the thread pool `self._cancel_pool`.
            - Returns the result of `self.cancel()`

********AsyncBaseExecutor********

- [ ] Business logic for communicating with the Runner:
    - `def _init_runtime(self, loop = None, cancel_pool: ThreadPoolExecutor = None)`:
        - This method is invoked by the Runner in `_run_task` to initialize runtime attributes of the executor instance.
        - Creates the following instance attributes:
            - `self._send_queue: asyncio.Queue` for sending messages to the Runner
            - `self._recv_queue: asyncio.Queue` for receiving messages from the Runner
            
            There is no need to store the `loop` or `cancel_pool` in this case. The method signature preserves consistency with non-async executors.
            
    - `def _notify(self, action: str, body: str = None)`
        - This will be invoked from within the `execute()` method, which runs in the server’s event loop.
        - Puts the message `(action, body)` in `self._send_queue`, return immediately (use `asyncio.Queue.put_nowait`).
        - In practice, `body` will be a JSON string.
    - `async def _wait_for_response(self, timeout: int = 5)`
        - Blocks the calling coroutine for up to `timeout` seconds for a response in `self._recv_queue`. and raise a `Empty` exception if no response is received. Use `asyncio.wait_for`. The response will be of the form `(status, body)`. where `status` is `True|False`.
        - Raise `RuntimeError` if `status` is False, indicating some sort of failure.
    - `async def _notify_sync(self, action: str, body: str = None)`
        - Convenience method to call `self._notify` and then block the calling coroutine using `self._wait_for_response`
- [ ] Methods for consumption by executor plugin authors:
    - `async def set_job_handle(self, job_handle)`
        - Saves a job handle (such as a slurm job id or Braket jobArn) to the database.
        - `job_handle` is required to be JSON-serializable — typically a string or int
        - Invokes `self._notify_sync()` with the following message
            - action: `"put"`
            - body: `("job_handle", json.dumps(job_handle))`
    - `async def get_cancel_requested(self)`
        - Checks whether the user has requested to cancel the job being processed by the executor instance.
        - Invokes `self._notify_sync()` with the following message
            - action: `"get"`
            - body: `"cancel_requested"`
- [ ] Modifications to `execute()`:
    - The return value of `execute()` is now `(output, stdout, stderr, status)`, where `status` is the job’s terminal status. The possible values are:
        - `RESULT_STATUS.COMPLETED`
        - `RESULT_STATUS.FAILED`
        - `RESULT_STATUS.CANCELLED`
    - In `execute()`, catch `TaskCancelledError` from `run()` and set the task’s terminal `job_status` to `RESULT_STATUS.CANCELLED`
    - The handler for `TaskRuntimeError` sets `status` to `RESULT_STATUS.FAILED`
    - Call `self._notify()` in the `finally:` block with the following message:
        - action: `"bye"`
        - body: `None`
        
        This tells the Runner to stop listening for messages from this executor instance.
        
- [ ] Methods to cancel jobs:
    - `async def cancel(self, task_metadata: Dict, job_handle: Any) -> bool`
        - To be overridden by subclasses with backend-specific logic to cancel the job with the specified `job_handle`
        - Return `True` if job cancellation succeeded and `False` otherwise
        - Default implementation should simply return `False` and log a warning that job cancellation is not supported for the specific executor type.
    - `async def _cancel(self, task_metadata: Dict, job_handle: Any) -> bool`
        - To be invoked by the Runner
            - Awaits `self.cancel()` followed by `self.teardown()`
            - Returns the result of `self.cancel()`

**DaskExecutor**

- [ ] In `run()`, save the Dask future's key using `Executor.set_job_handle()`
- [ ] Implement the `cancel()` method for the DaskExecutor:
    - Initialize a dask client
    - Create a Dask Future from the key
    - Await `Future.cancel()`
See [here](https://github.com/AgnostiqHQ/covalent/blob/cancel-dispatch-simplified/covalent/executor/executor_plugins/dask.py#L125) for example.

**Runner**

- [ ] Introduce **executor proxy** in `_core/runner_modules/executor_proxy.py` — Interacts with the Job Manager on the executor’s behalf
    - `async def watch(dispatch_id: str, task_id: int, executor: _AbstractBaseExecutor)`
        - `executor` is the plugin instance running the task `(dispatch_id, task_id)`
        - Listens for messages in `executor._send_queue` and invokes the appropriate handler
        - Puts replies in `executor._recv_queue` using `put.nowait()`
        - See message protocol below
    - Implement handlers for each of the messages below. **The handlers should deal with negative `task_id`s appropriately -- the postprocessing task with id `-1` does not exist in the database.**
    - 
- [ ] Modifications to `_run_task:`
    - After populating the executor with task-specific data, initialize the executor plugin’s runtime properties by calling `executor._init_runtime()` with the Runner’s event loop
    - Start `executor_proxy.watch` as a background asyncio task just before invoking `Executor._execute`. The implementation of `BaseExecutor._execute` sends a `"bye"` message right before returning to terminate the `watch` task. All references to the executor plugin are released automatically.
    - The return value of `Executor._execute`:  has changed from `(output, stdout, stderr, exception_raised)` to `(output, stdout, stderr, status)`. The generated node result can simply set `status=status`.
- [ ] Create a singleton `ThreadPoolExecutor` called `_cancel_pool` at the top of `[runner.py](http://runner.py)` . This will be reserved specifically for cancelling jobs running in non-async executors.
- [ ] In `_run_abstract_task`, query the Job Manager to see if the task has been requested to be cancelled. If that is so, **do not** proceed to `_run_task`, and simply use `data_manager.generate_node_result()` to return a cancelled node result:
    
    ```python
    {
    	"node_id": node_id, 
    	"start_time": timestamp, 
    	"end_time": timestamp, 
    	"status": RESULT_STATUS.CANCELLED
    }
    
    ```
    
- [ ] `async def _cancel_task(dispatch_id: str, task_id: int, executor: str, executor_data: Dict, job_handle: str) -> None`
    - Params:
        - `executor` is the short name of the task’s executor, such as `"dask"`, `"local"`, etc.
        - `executor_data` is the dictionary of executor attributes — to be set via `Executor.from_dict()` (see how `_run_task` initializes executors)
        - `job_handle` is the JSON-serialized job handle (see `BaseExecutor.set_job_handle`) saved by the executor plugin
    - Effect:
        - Initializes a new executor instance of type `executor` and attributes `executor_data` (see how this is done in `_run_task`)
        - Passes the running event loop and `_cancel_pool` to `Executor._init_runtime()`
        - Awaits `executor._cancel(task_metadata, json.loads(job_handle)` where `task_metadata = {"dispatch_id": dispatch_id, "node_id": task_id}`
        - Sets the job’s `cancel_result` to the return value of `executor._cancel`  (use Job Manager)
        - If any exception is raised during this entire process, set `cancel_result` to `False`.
- [ ] `async def cancel_tasks(dispatch_id: str, task_ids: List[int])`
    - Cancels the batch of tasks specified by `task_ids`. For each task, invoke `_cancel_task()`  in a background asyncio task.

- [ ] Add and adjust unit tests. Feel free to crib some tests from [this branch](https://github.com/AgnostiqHQ/covalent/tree/cancel-dispatch-simplified/tests). Mock all interactions with the other core modules.

### LINKED ISSUE #1473: Cancelling dispatches: Dispatcher and API
As part of the [epic to support cancelling tasks and dispatches](https://github.com/AgnostiqHQ/covalent-staging/issues/662), this issue tracks the proposed enhancements to the Dispatcher module and API.

## Acceptance Criteria

**Dispatcher**

- [ ] `async def cancel_dispatch(dispatch_id: str, task_ids: List[int])`
    - Effect:
        - Let `task_ids_to_cancel = task_ids` if `task_ids` is nonempty, otherwise set `task_ids_to_cancel` to all tasks in the workflow.
        - **Await** `job_manager.set_cancel_requested` on the batch of tasks; after that returns, call `runner.cancel_tasks`.
        - For each electron in `task_ids_to_cancel` that represents a running sublattice (indicated by a nonempty `sub_dispatch_id` property), recursively cancel the sublattice dispatch.

**Public API:**

- [ ] Implement the server endpoint: In `_service/app.py`, implement the `"/cancel"` endpoint:
    - JSON request body:
        
        ```python
        {
        	"dispatch_id": str,
        	"task_ids": List[int],
        }
        ```
        
    - Sample Responses:
        - If `task_ids` is nonempty:
            - `f“Cancelling tasks {task_ids} in dispatch {dispatch_id}”`
        - If `task_ids` is an empty list:
            - `f”Cancelling dispatch {dispatch_id}”`
- [ ] SDK: Implement in `_results_manager/results_manager.py`:
    - `def cancel(dispatch_id: str, task_ids: List[int] = [], dispatcher_addr: str = None)`
        - See [the PR](https://github.com/AgnostiqHQ/covalent/pull/1460) for `get_result` for how to determine the server URL.

- [ ] Add comprehensive unit tests. Feel free to crib some from [this branch](https://github.com/AgnostiqHQ/covalent/tree/cancel-dispatch-simplified/tests).

### LINKED ISSUE #1474: Cancelling dispatches: end-to-end tests
This issue rounds out the [epic to support cancelling tasks and dispatches](https://github.com/AgnostiqHQ/covalent-staging/issues/662).  As this feature involves several moving parts, it is important to verify through functional tests that they interact correctly:

### Acceptance criteria

- [ ] Test cancelling an entire dispatch
- [ ] Test cancelling specific tasks within a dispatch. Verify that tasks without a cancelled parent complete successfully.
- [ ] Test cancelling a task in a dispatch that corresponds to a sublattice. Give the sublattice dispatch time to start running. Verify via `ct.get_result(sub_dispatch_id)` that the sublattice dispatch status is cancelled.

- [ ] Please incorporate at least the [linked functional tests.](https://github.com/AgnostiqHQ/covalent/blob/cancel-dispatch-simplified/tests/functional_tests/workflow_cancellation_test.py)
- [ ] Check that the UI marks cancelled tasks as cancelled.
- [ ]  Document `ct.cancel()` in the RTD.

### PR DESCRIPTION
**Pull request by kessler-frost**
- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation and CHANGELOG accordingly.
- [ ] I have read the CONTRIBUTING document.

Resolves #1471 
Closes #1162 
closes #1474 
closes #1472 
closes #1473 
Closes #1471

### CODE REVIEW

File: covalent/_workflow/transport.py [hunk 1]
```diff
@@ -312,6 +312,10 @@ def __init__(self) -> None:
             "stderr": None,
         }
 
+    def add_node_by_id(self, node_id: int, **attr):
```
Comment (cjao) [inline covalent/_workflow/transport.py:315]: Is this used anywhere?

**Review by cjao — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 2]
```diff
@@ -7,9 +7,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [UNRELEASED]
 
-### Changed
+### Added
 
-- Enhanced the Dockerfile to include builds from various sources and a differentiation between SDK and Server builds
+- Adding more tests to bring up code coverage
```
[the remark this answers is not in the document]
Comment (venkatBala) [reply CHANGELOG.md:12]: @cjao haha true... I'll add more detailed updates to the changelog

File: covalent/_workflow/transport.py (hunk 1 above)
Comment (venkatBala) [reply covalent/_workflow/transport.py:315]: @cjao I believe it was... I'll double check

File: covalent_dispatcher/_service/app.py [hunk 3]
```diff
@@ -103,17 +207,23 @@ async def cancel(request: Request) -> str:
...
+    else:
         return f"Dispatch {dispatch_id} cancelled."
-    except Exception as e:
-        raise HTTPException(
-            status_code=400,
-            detail=f"Failed to cancel workflow: {e}",
-        ) from e
+
+
+@router.get("/db-path")
```
Comment (AlejandroEsquivel) [inline covalent_dispatcher/_service/app.py:223]: @venkatBala I think this got removed in another PR, let's get rid of this endpoint please

File: covalent/executor/base.py [hunk 4]
```diff
@@ -202,6 +206,31 @@ def __init__(
...
+        return self._wait_for_response()
+
+    def _wait_for_response(self, timeout: int = 5):
+        status, body = self._recv_queue.get(timeout=timeout)
+        if status is None:
+            raise RuntimeError("Error waiting for response")
+        return body
+
+    def get_cancel_requested(self):
+        return self._notify_sync("get", "cancel_requested")
```
Comment (AlejandroEsquivel) [inline covalent/executor/base.py:229]: @venkatBala let's try to use constants or enums here instead of string literals

**Review by AlejandroEsquivel — CHANGES_REQUESTED** [review]
Looks great @venkatBala just some minor comments

File: covalent/executor/base.py (hunk 4 above)
Comment (venkatBala) [reply covalent/executor/base.py:229]: @AlejandroEsquivel I understand... but the implementation has been as such that the threads on the receiving end of the queue are expecting these string literals to process them in task cancellation

File: covalent_dispatcher/_service/app.py (hunk 3 above)
Comment (venkatBala) [reply covalent_dispatcher/_service/app.py:223]: removed

File: tests/covalent_dispatcher_tests/_core/data_modules/job_manager_test.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    get_jobs_metadata,
+    set_cancel_requested,
+    set_cancel_result,
+    set_job_handle,
+)
+
+
+def get_mock_tg():
+    tg = _TransportGraph()
+    tg.add_node_by_id(0, job_id=1)
```
Comment (cjao) [inline tests/covalent_dispatcher_tests/_core/data_modules/job_manager_test.py:36]: it is not necessary to create a new `TransportGraph` method just for this test. `TransportGraph.add_node()` should suffice.

**Review by cjao — CHANGES_REQUESTED** [review]

File: tests/covalent_dispatcher_tests/_core/data_modules/job_manager_test.py (hunk 5 above)
Comment (cjao) [reply tests/covalent_dispatcher_tests/_core/data_modules/job_manager_test.py:36]: For more context, `TransportGraph.add_node_by_id()` was initially added to support cancelling postprocessing -- but that is no longer necessary in view of upcoming changes to how postprocessing will handled.

Comment (venkatBala) [reply tests/covalent_dispatcher_tests/_core/data_modules/job_manager_test.py:36]: @cjao ah I see... alright will deprecate this method and update the affected tests

**Review by AlejandroEsquivel — APPROVED** [review]

**Review by cjao — APPROVED** [review]

**Pull request merged**